maas/client/viscera/boot_resources.py

```diff
@@ -0,0 +1,285 @@
+"""Objects for boot resources."""
+
+__all__ = [
+    "BootResource",
+    "BootResources",
+]
+
+import enum
+import hashlib
+import httplib2
+import io
+from typing import Callable
+
+from . import (
+    check,
+    check_optional,
+    mapping_of,
+    Object,
+    ObjectField,
+    ObjectSet,
+    ObjectType,
+)
+from .. import utils
+from ..bones import CallError
+
+
+def calc_size_and_sha265(content: io.IOBase, chunk_size: int):
+    """Calculates the size and the sha2566 value of the content."""
+    size = 0
+    sha256 = hashlib.sha256()
+    content.seek(0, io.SEEK_SET)
+    while True:
+        buf = content.read(chunk_size)
+        length = len(buf)
+        size += length
+        sha256.update(buf)
+        if length != chunk_size:
+            break
```

Comment:
EOF is when `length` is zero. From `io.BufferedReader.read` (which is what you get when opening a file):

> ... for interactive raw streams (as well as sockets and pipes),
> at most one raw read will be issued, and a short result does
> not imply that EOF is imminent.
> 
> **Returns an empty bytes object on EOF.**